content/blog/kubecon-2019/index.md

```diff
@@ -0,0 +1,272 @@
+
+We spin up our Kubernetes Job and additional resources by using an if statement at the bottom of our `infrastructure/index.ts` file. The conditional depends on the environment's name as follows:
+
+```typescript
+// If it's a test environment, set up acceptance tests.
+let job: k8s.batch.v1.Job | undefined;
+if (ENV.startsWith("test")) {
+    job = acceptance.setupAcceptanceTests({
+        ...
+    });
+}
+
+// Export the acceptance job name, so we can get the logs from our
+// acceptance tests.
+export const acceptanceJobName = job ? job.metadata.name : "unapplicable";
+```
+
+That covers all the major aspects of our application and infrastructure, and if you'd like to view the code in detail it is all available for reference in our `demo-app` [GitLab repository](https://gitlab.com/rocore/demo-app).
+
+## Our Pipeline
+
+When developing a new service, it’s important that we establish a solid deployment strategy upfront. We want to make sure we're building in quality from day one. As we develop the service, we can add acceptance tests for each and every feature we add while the context and requirements are still fresh in our minds. This ensures we have thorough coverage of our app's functionality.
+
+We used GitLab to set up our pipeline. We chose GitLab because its extremely easy to set up and allows us to run our pipeline on our Docker image of choice. We use a [base-image](https://gitlab.com/rocore/global-infra/blob/master/base-image/Dockerfile) that has all our dependencies installed and then reference that Docker image and tag in our `demo-app` pipeline. This allows us to bundle and version the dependencies for building our application and infrastructure.
+
+![gitlab ci pipeline](gitlab-ci.png)
+
+1. **Test and Build** - This runs our unit tests and builds both our application and acceptance test images. The application image is an immutable image that is used for both running our acceptance tests and deploying to production.
+1. **Acceptance Test** - This is what spins up our ephemeral environments and runs our acceptance tests. This acts a quality gate catching issues prior to production.
+
+    Our ephemeral environment and Kubernetes job are all spun up in the `script`.
+
+    ```bash
+    ...
+    $ pulumi stack init rocore/$ENV-app
+    Logging in using access token from PULUMI_ACCESS_TOKEN
+    Created stack 'rocore/test-96425413-app'
+    $ pulumi config set DOCKER_TAG $DOCKER_TAG
+    $ pulumi config set ENV $ENV
+    $ pulumi config set gcp:project rocore-k8s
+    $ pulumi config set gcp:zone us-west1-a
+    $ pulumi up --skip-preview
+    Updating (rocore/test-96425413-app):
+    ...
+    Resources:
+        + 16 created
+
+    Duration: 4m10s
+
+    Permalink: https://app.pulumi.com/rocore/demo-app/test-96425413-app/updates/1
+    ```
+
+    The after script destroys our stack as well as prints the logs of both our Kubernetes Job and Deployment, which are helpful for debugging if our tests were to fail.
```

Comment:
"codify" after script with back ticks, only because we did it when we said script.

Also, maybe call out that after-script runs either way, whether the job succeeds or fails, so our ephemeral resources are always  cleaned up. Not sure if this is necessary though.